cpp/src/arrow/flight/client.cc

```diff
@@ -993,6 +1010,36 @@ class FlightClient::FlightClientImpl {
     return Status::OK();
   }
 
+  Status AuthenticateBasicToken(const std::string& username, const std::string& password,
+                                std::pair<std::string, std::string>* bearer_token) {
+    // Add bearer token factory to middleware so it can intercept the bearer token.
+    if (interceptor_pointer != NULLPTR) {
+      interceptor_pointer->AddMiddlewareFactory(
+          std::make_shared<internal::ClientBearerTokenFactory>(bearer_token));
+    } else {
+      return MakeFlightError(FlightStatusCode::Internal,
+                             "Connect must be called before AuthenticateBasicToken.");
+    }
+    ClientRpc rpc({});
```

Comment:
You may actually want to allow passing call options so things like timeouts can be set.